calyx-backend/src/firrtl.rs

```diff
@@ -0,0 +1,181 @@
+fn emit_component<F: io::Write>(
+    comp: &ir::Component,
+    f: &mut F,
+) -> io::Result<()> {
+    writeln!(f, "circuit {}:", comp.name)?;
+    writeln!(f, "{}module {}:", SPACING, comp.name)?;
+
+    // Inputs and Outputs
+    let sig = comp.signature.borrow();
+    for (_idx, port_ref) in sig.ports.iter().enumerate() {
+        let port = port_ref.borrow();
+        let direction_string =
+        // NOTE: The signature port definitions are reversed inside the component.
+        match port.direction {
+            ir::Direction::Input => {"output"}
+            ir::Direction::Output => {"input"}
+            ir::Direction::Inout => {
+                panic!("Unexpected Inout port on Component: {}", port.name)
+            }
+        };
+        // FIXME: Hack to get clock declaration right. Should check for attribute name instead.
+        if port.name == "clk" {
+            writeln!(
+                f,
+                "{}{} {}: Clock",
+                SPACING.repeat(2),
+                direction_string,
+                port.name
+            )?;
+        } else {
+            writeln!(
+                f,
+                "{}{} {}: UInt<{}>",
+                SPACING.repeat(2),
+                direction_string,
+                port.name,
+                port.width
+            )?;
+        }
+    }
+
+    // Add a COMPONENT START: <name> anchor before any code in the component
+    writeln!(f, "{}; COMPONENT START: {}", SPACING.repeat(2), comp.name)?;
+
+    // TODO: Cells. NOTE: leaving this one for last
+
+    for asgn in &comp.continuous_assignments {
+        // TODO: guards
+        match asgn.guard.as_ref() {
+            ir::Guard::Or(_, _) => todo!(),
+            ir::Guard::And(_, _) => todo!(),
+            ir::Guard::Not(_) => todo!(),
+            ir::Guard::True => {
+                // Simple assignment with no guard
+                let _ = write_assignment(asgn, f);
+            }
+            ir::Guard::CompOp(_, _, _) => todo!(),
+            ir::Guard::Port(_) => {}
+            ir::Guard::Info(_) => todo!(),
+        }
+    }
+
+    // Add COMPONENT END: <name> anchor
+    writeln!(f, "{}; COMPONENT END: {}", SPACING.repeat(2), comp.name)?;
+
+    Ok(())
+}
+
+fn write_assignment<F: io::Write>(
+    asgn: &ir::Assignment<ir::Nothing>,
+    f: &mut F,
+) -> CalyxResult<()> {
+    let dest_port = asgn.dst.borrow();
+    let mut dest_string = SPACING.repeat(2);
+    match &dest_port.parent {
+        ir::PortParent::Cell(cell) => {
+            let parent_ref = cell.upgrade();
+            let parent = parent_ref.borrow();
+            match parent.prototype {
+                ir::CellType::ThisComponent => {
+                    dest_string.push_str(dest_port.name.as_ref());
+                }
+                _ => {
+                    let formatted = format!(
+                        "{}.{}",
+                        parent.name().as_ref(),
+                        dest_port.name.as_ref()
```

Comment:
These two `as_ref()`s can be removed since the `Id` type implements `Display`
which is all that `format!` needs.